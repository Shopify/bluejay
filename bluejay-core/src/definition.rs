mod arguments_definition;
mod directive_definition;
mod enum_type_definition;
mod enum_value_definition;
mod enum_value_definitions;
mod field_definition;
mod fields_definition;
mod has_directives;
mod input_fields_definition;
mod input_object_type_definition;
mod input_type;
mod input_value_definition;
mod interface_implementation;
mod interface_implementations;
mod interface_type_definition;
mod object_type_definition;
mod output_type;
mod scalar_type_definition;
mod schema_definition;
mod type_definition;
mod union_member_type;
mod union_member_types;
mod union_type_definition;

pub use arguments_definition::ArgumentsDefinition;
pub use directive_definition::{DirectiveDefinition, DirectiveLocation};
pub use enum_type_definition::EnumTypeDefinition;
pub use enum_value_definition::EnumValueDefinition;
pub use enum_value_definitions::EnumValueDefinitions;
pub use field_definition::FieldDefinition;
pub use fields_definition::FieldsDefinition;
pub use has_directives::HasDirectives;
pub use input_fields_definition::InputFieldsDefinition;
pub use input_object_type_definition::InputObjectTypeDefinition;
pub use input_type::{BaseInputTypeReference, InputType, InputTypeReference};
pub use input_value_definition::InputValueDefinition;
pub use interface_implementation::InterfaceImplementation;
pub use interface_implementations::InterfaceImplementations;
pub use interface_type_definition::InterfaceTypeDefinition;
pub use object_type_definition::ObjectTypeDefinition;
pub use output_type::{BaseOutputTypeReference, OutputType, OutputTypeReference};
pub use scalar_type_definition::ScalarTypeDefinition;
pub use schema_definition::SchemaDefinition;
pub use type_definition::{TypeDefinition, TypeDefinitionReference};
pub use union_member_type::UnionMemberType;
pub use union_member_types::UnionMemberTypes;
pub use union_type_definition::UnionTypeDefinition;

pub mod prelude {
    pub use super::{
        ArgumentsDefinition as _, DirectiveDefinition as _, EnumTypeDefinition as _,
        EnumValueDefinition as _, EnumValueDefinitions as _, FieldDefinition as _,
        FieldsDefinition as _, HasDirectives as _, InputFieldsDefinition as _,
        InputObjectTypeDefinition as _, InputType as _, InputValueDefinition as _,
        InterfaceImplementation as _, InterfaceImplementations as _, InterfaceTypeDefinition as _,
        ObjectTypeDefinition as _, OutputType as _, ScalarTypeDefinition as _,
        SchemaDefinition as _, TypeDefinition as _, UnionMemberType as _, UnionMemberTypes as _,
        UnionTypeDefinition as _,
    };
}
