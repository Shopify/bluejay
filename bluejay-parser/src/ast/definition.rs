mod arguments_definition;
mod context;
mod custom_scalar_type_definition;
mod definition_document;
mod directive_definition;
mod enum_type_definition;
mod enum_value_definition;
mod enum_value_definitions;
mod explicit_schema_definition;
mod field_definition;
mod fields_definition;
mod input_fields_definition;
mod input_object_type_definition;
mod input_type;
mod input_value_definition;
mod interface_implementation;
mod interface_implementations;
mod interface_type_definition;
mod object_type_definition;
mod output_type_reference;
mod schema_definition;
mod type_definition_reference;
mod union_member_type;
mod union_member_types;
mod union_type_definition;

pub use arguments_definition::ArgumentsDefinition;
pub use context::{Context, DefaultContext};
pub use custom_scalar_type_definition::CustomScalarTypeDefinition;
pub use definition_document::DefinitionDocument;
pub use directive_definition::DirectiveDefinition;
pub use enum_type_definition::EnumTypeDefinition;
pub use enum_value_definition::EnumValueDefinition;
pub use enum_value_definitions::EnumValueDefinitions;
pub use explicit_schema_definition::{ExplicitSchemaDefinition, RootOperationTypeDefinition};
pub use field_definition::FieldDefinition;
pub use fields_definition::FieldsDefinition;
pub use input_fields_definition::InputFieldsDefinition;
pub use input_object_type_definition::InputObjectTypeDefinition;
pub use input_type::{BaseInputType, InputType};
pub use input_value_definition::InputValueDefinition;
pub use interface_implementation::InterfaceImplementation;
pub use interface_implementations::InterfaceImplementations;
pub use interface_type_definition::InterfaceTypeDefinition;
pub use object_type_definition::ObjectTypeDefinition;
pub use output_type_reference::{BaseOutputType, OutputTypeReference};
pub use schema_definition::SchemaDefinition;
pub use type_definition_reference::TypeDefinitionReference;
pub use union_member_type::UnionMemberType;
pub use union_member_types::UnionMemberTypes;
pub use union_type_definition::UnionTypeDefinition;
