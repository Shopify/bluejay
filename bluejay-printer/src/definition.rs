mod arguments_definition;
mod directive_definition;
mod enum_type_definition;
mod field_definition;
mod input_object_type_definition;
mod input_value_definition;
mod interface_implementations;
mod interface_type_definition;
mod object_type_definition;
mod scalar_type_definition;
mod schema_definition;
mod union_type_definition;

pub use schema_definition::SchemaDefinitionPrinter;
