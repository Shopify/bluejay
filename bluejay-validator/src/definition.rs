mod error;
mod rule;
mod rules;
mod validator;
mod visitor;

pub use error::Error;
pub use rule::Rule;
pub use rules::BuiltinRules;
pub use validator::{BuiltinRulesValidator, Validator};
pub use visitor::Visitor;
