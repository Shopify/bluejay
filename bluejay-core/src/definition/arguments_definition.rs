use crate::definition::InputValueDefinition;
use crate::AsIter;

pub trait ArgumentsDefinition: AsIter<Item=Self::ArgumentDefinition> {
    type ArgumentDefinition: InputValueDefinition;
}
