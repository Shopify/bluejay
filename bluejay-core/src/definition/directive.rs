use crate::definition::DirectiveDefinition;

pub trait Directive: crate::Directive<true> {
    type DirectiveDefinition: DirectiveDefinition;

    fn definition(&self) -> &Self::DirectiveDefinition;
}

pub trait Directives: crate::Directives<true, Directive = <Self as Directives>::Directive> {
    type Directive: Directive;
}
