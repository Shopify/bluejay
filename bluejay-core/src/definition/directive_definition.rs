use crate::definition::ArgumentsDefinition;
use crate::AsIter;
use strum::{AsRefStr, Display, EnumIter, EnumString, EnumVariantNames, VariantNames};

#[derive(
    Debug, Clone, Copy, EnumString, EnumVariantNames, EnumIter, AsRefStr, Display, PartialEq, Eq,
)]
#[strum(serialize_all = "SCREAMING_SNAKE_CASE")]
pub enum DirectiveLocation {
    Query,
    Mutation,
    Subscription,
    Field,
    FragmentDefinition,
    FragmentSpread,
    InlineFragment,
    VariableDefinition,
    Schema,
    Scalar,
    Object,
    FieldDefinition,
    ArgumentDefinition,
    Interface,
    Union,
    Enum,
    EnumValue,
    InputObject,
    InputFieldDefinition,
}

impl DirectiveLocation {
    pub const POSSIBLE_VALUES: &'static [&'static str] = Self::VARIANTS;
}

pub trait DirectiveDefinition {
    type ArgumentsDefinition: ArgumentsDefinition;
    type DirectiveLocations: AsIter<Item = DirectiveLocation>;

    fn description(&self) -> Option<&str>;
    fn name(&self) -> &str;
    fn arguments_definition(&self) -> Option<&Self::ArgumentsDefinition>;
    fn is_repeatable(&self) -> bool;
    fn locations(&self) -> &Self::DirectiveLocations;
    fn is_builtin(&self) -> bool;
}
