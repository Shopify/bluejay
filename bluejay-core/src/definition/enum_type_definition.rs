use crate::definition::EnumValueDefinitions;

pub trait EnumTypeDefinition {
    type EnumValueDefinitions: EnumValueDefinitions;

    fn description(&self) -> Option<&str>;
    fn name(&self) -> &str;
    fn enum_value_definitions(&self) -> &Self::EnumValueDefinitions;
}
