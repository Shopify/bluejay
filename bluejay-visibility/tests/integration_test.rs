use bluejay_core::{
    definition::{prelude::*, SchemaDefinition as CoreSchemaDefinition, TypeDefinitionReference},
    AsIter, Directive as _,
};
use bluejay_parser::{
    ast::{
        definition::{
            DefaultContext, DefinitionDocument, Directives,
            SchemaDefinition as ParserSchemaDefinition,
        },
        Parse,
    },
    Error,
};
use bluejay_printer::definition::SchemaDefinitionPrinter;
use bluejay_visibility::{Cache, SchemaDefinition, Warden};
use std::marker::PhantomData;

#[derive(Default)]
struct DirectiveWarden<'a>(PhantomData<ParserSchemaDefinition<'a>>);

impl<'a> DirectiveWarden<'a> {
    fn has_visible_directive(directives: Option<&Directives<'a, DefaultContext>>) -> bool {
        directives.map_or(false, |directives| {
            directives
                .iter()
                .any(|directive| directive.name() == "visible")
        })
    }
}

impl<'a> Warden for DirectiveWarden<'a> {
    type SchemaDefinition = ParserSchemaDefinition<'a>;
    type TypeDefinitionsForName<'b> = std::option::IntoIter<
        TypeDefinitionReference<
            'b,
            <Self::SchemaDefinition as CoreSchemaDefinition>::TypeDefinition,
        >,
    > where Self: 'b;

    fn is_enum_value_definition_visible(
        &self,
        enum_value_definition: &<Self::SchemaDefinition as bluejay_core::definition::SchemaDefinition>::EnumValueDefinition,
    ) -> bool {
        Self::has_visible_directive(enum_value_definition.directives())
    }

    fn is_field_definition_visible(
        &self,
        field_definition: &<Self::SchemaDefinition as bluejay_core::definition::SchemaDefinition>::FieldDefinition,
    ) -> bool {
        Self::has_visible_directive(field_definition.directives())
    }

    fn is_input_value_definition_visible(
        &self,
        input_value_definition: &<Self::SchemaDefinition as bluejay_core::definition::SchemaDefinition>::InputValueDefinition,
    ) -> bool {
        Self::has_visible_directive(input_value_definition.directives())
    }

    fn is_interface_implementation_visible(
        &self,
        _: &<Self::SchemaDefinition as bluejay_core::definition::SchemaDefinition>::InterfaceImplementation,
    ) -> bool {
        true
    }

    fn is_union_member_type_visible(
        &self,
        _: &<Self::SchemaDefinition as bluejay_core::definition::SchemaDefinition>::UnionMemberType,
    ) -> bool {
        true
    }

    fn is_directive_definition_visible(
        &self,
        _: &<Self::SchemaDefinition as bluejay_core::definition::SchemaDefinition>::DirectiveDefinition,
    ) -> bool {
        true
    }

    fn is_custom_scalar_type_definition_visible(
        &self,
        custom_scalar_type_definition: &<Self::SchemaDefinition as bluejay_core::definition::SchemaDefinition>::CustomScalarTypeDefinition,
    ) -> bool {
        Self::has_visible_directive(custom_scalar_type_definition.directives())
    }

    fn is_enum_type_definition_visible(
        &self,
        enum_type_definition: &<Self::SchemaDefinition as bluejay_core::definition::SchemaDefinition>::EnumTypeDefinition,
    ) -> bool {
        Self::has_visible_directive(enum_type_definition.directives())
    }

    fn is_input_object_type_definition_visible(
        &self,
        input_object_type_definition: &<Self::SchemaDefinition as bluejay_core::definition::SchemaDefinition>::InputObjectTypeDefinition,
    ) -> bool {
        Self::has_visible_directive(input_object_type_definition.directives())
    }

    fn is_interface_type_definition_visible(
        &self,
        interface_type_definition: &<Self::SchemaDefinition as bluejay_core::definition::SchemaDefinition>::InterfaceTypeDefinition,
    ) -> bool {
        Self::has_visible_directive(interface_type_definition.directives())
    }

    fn is_object_type_definition_visible(
        &self,
        object_type_definition: &<Self::SchemaDefinition as bluejay_core::definition::SchemaDefinition>::ObjectTypeDefinition,
    ) -> bool {
        Self::has_visible_directive(object_type_definition.directives())
    }

    fn is_union_type_definition_visible(
        &self,
        union_type_definition: &<Self::SchemaDefinition as bluejay_core::definition::SchemaDefinition>::UnionTypeDefinition,
    ) -> bool {
        Self::has_visible_directive(union_type_definition.directives())
    }

    fn type_definitions_for_name<'b>(
        &self,
        schema_definition: &'b Self::SchemaDefinition,
        type_name: &str,
    ) -> Self::TypeDefinitionsForName<'b> {
        schema_definition.get_type_definition(type_name).into_iter()
    }
}

#[test]
fn test_visibility() {
    insta::glob!("test_data/*.graphql", |path| {
        let input = std::fs::read_to_string(path).unwrap();
        let definition_document: DefinitionDocument = DefinitionDocument::parse(&input)
            .unwrap_or_else(|errors| {
                panic!(
                    "Schema `{}` had parse errors:\n{}",
                    path.display(),
                    Error::format_errors(&input, path.file_name().and_then(|f| f.to_str()), errors)
                )
            });
        let schema_definition = ParserSchemaDefinition::try_from(&definition_document)
            .unwrap_or_else(|errors| {
                panic!(
                    "Schema `{}` had coercion errors:\n:{}",
                    path.display(),
                    Error::format_errors(&input, path.file_name().and_then(|f| f.to_str()), errors)
                )
            });

        let cache = Cache::new(DirectiveWarden::default(), &schema_definition);
        let visibility_scoped_schema_definition = SchemaDefinition::new(&cache).unwrap();

        let printed_schema_definition =
            SchemaDefinitionPrinter::to_string(&visibility_scoped_schema_definition);

        insta::assert_snapshot!(printed_schema_definition);
    });
}

#[test]
fn test_fields_definition_get() {
    let schema = "
        directive @visible on FIELD_DEFINITION | ENUM_VALUE | INPUT_FIELD_DEFINITION | ARGUMENT_DEFINITION | SCALAR | OBJECT | INTERFACE | UNION | ENUM

        type Query @visible {
            otherField: String @visible
            field: String
            field: String @visible
        }
    ";

    let definition_document: DefinitionDocument =
        DefinitionDocument::parse(schema).unwrap_or_else(|errors| {
            panic!(
                "Schema had parse errors:\n{}",
                Error::format_errors(schema, None, errors)
            )
        });
    let schema_definition =
        ParserSchemaDefinition::try_from(&definition_document).unwrap_or_else(|errors| {
            panic!(
                "Schema had coercion errors:\n:{}",
                Error::format_errors(schema, None, errors)
            )
        });

    let cache = Cache::new(DirectiveWarden::default(), &schema_definition);
    let visibility_scoped_schema_definition = SchemaDefinition::new(&cache).unwrap();

    assert_eq!(
        visibility_scoped_schema_definition
            .query()
            .fields_definition()
            .get("field")
            .map(|fd| fd.name()),
        Some("field"),
        "Expected field to be visible",
    );
}
