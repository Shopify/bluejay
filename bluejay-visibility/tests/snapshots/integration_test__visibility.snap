---
source: bluejay-visibility/tests/integration_test.rs
expression: printed_schema_definition
---
directive @visible on FIELD_DEFINITION | ENUM_VALUE | INPUT_FIELD_DEFINITION | ARGUMENT_DEFINITION | SCALAR | OBJECT | INTERFACE | UNION | ENUM

enum MyEnum @visible {
  VISIBLE @visible
}

input MyInput @visible {
  visible: Int @visible
}

type MyObject @visible {
  visible(
    visible: Int @visible
  ): Int @visible
}

scalar MyScalar @visible

type Query @visible {
  myEnum: MyEnum @visible

  myObject: MyObject @visible

  myScalar: MyScalar @visible
}

